pkg/actors/internal/client.go

```diff
@@ -15,162 +15,108 @@ package internal
 
 import (
 	"context"
+	"errors"
 	"sync"
+	"sync/atomic"
 
 	v1pb "github.com/dapr/dapr/pkg/proto/placement/v1"
 
 	"google.golang.org/grpc"
 )
 
+type client interface {
+	Send(*v1pb.Host) error
+	Recv() (*v1pb.PlacementOrder, error)
+	Close()
+}
+
+type grpcOptionsFn func() ([]grpc.DialOption, error)
+
 // placementClient implements the best practices when handling grpc streams
 // using exclusion locks to prevent concurrent Recv and Send from channels.
-// properly handle stream closing by draining the renamining events until receives an error.
-// and broadcasts connection results based on new connections and disconnects.
+// properly handle stream closing by draining the renamining events until
+// receives an error.
 type placementClient struct {
-	// getGrpcOpts are the options that should be used to connect to the placement service
-	getGrpcOpts func() ([]grpc.DialOption, error)
-
 	// recvLock prevents concurrent access to the Recv() method.
-	recvLock *sync.Mutex
+	recvLock sync.Mutex
+
 	// sendLock prevents CloseSend and Send being used concurrently.
-	sendLock *sync.Mutex
-
-	// clientStream is the client side stream.
-	clientStream v1pb.Placement_ReportDaprStatusClient
-	// clientConn is the gRPC client connection.
-	clientConn *grpc.ClientConn
-	// streamConnAlive is the status of stream connection alive.
-	streamConnAlive bool
-	// streamConnectedCond is the condition variable for goroutines waiting for or announcing
-	// that the stream between runtime and placement is connected.
-	streamConnectedCond *sync.Cond
-
-	// ctx is the stream context
-	ctx context.Context
-	// cancel is the cancel func for context
-	cancel context.CancelFunc
+	// Sends use the read lock. CloseSend uses the write lock.
+	sendLock sync.RWMutex
+
+	// stream is the client side stream.
+	stream v1pb.Placement_ReportDaprStatusClient
+
+	// conn is the gRPC client connection.
+	conn *grpc.ClientConn
+
+	closed atomic.Bool
 }
 
-// connectToServer initializes a new connection to the target server and if it succeeds replace the current
-// stream with the connected stream.
-func (c *placementClient) connectToServer(serverAddr string) error {
-	opts, err := c.getGrpcOpts()
+func newPlacementClient(ctx context.Context, addr string, getGrpcOpts grpcOptionsFn) (client, error) {
+	opts, err := getGrpcOpts()
 	if err != nil {
-		return err
+		return nil, err
 	}
 
-	conn, err := grpc.Dial(serverAddr, opts...)
+	conn, err := grpc.DialContext(ctx, addr, opts...)
 	if err != nil {
 		if conn != nil {
 			conn.Close()
 		}
-		return err
+		return nil, err
 	}
 
-	ctx, cancel := context.WithCancel(context.Background())
-	client := v1pb.NewPlacementClient(conn)
-	stream, err := client.ReportDaprStatus(ctx)
+	stream, err := v1pb.NewPlacementClient(conn).ReportDaprStatus(ctx)
 	if err != nil {
-		cancel()
-		return err
+		return nil, err
```

Comment:
This should fix #6130

```suggestion
        if conn != nil {
			conn.Close()
		}
		return nil, err
```